Pass experimental_relax_shapes to instance methods decorated with `tf.function`.

foxik opened the pull request:
Currently the `experimental_relax_shapes` is not passed, so the instance methods ignore this argument.

Fixes #34905.

BTW, there are other options not passed, like experimental_autograph_options. However, in `eager.py:Function` the argument is `experimental_autograph_options` while in `function.py:Function` it is called just `autograph_options`, and I am not sure if the `class_method_to_instance_method` cannot be used for both. Therefore this pull request handles only `experimental_relax_shapes` which has the same name everywhere.

foxik committed:
Pass experimental_relax_shapes to instance methods...

decorated with `tf.function`.

Currently the `experimental_relax_shapes` is not passed,
so the instance methods ignore this argument.

Fixes #34905.

Note that the `experimental_relax_shapes` was named differently in
`def_function.Function` and `function.Function`, so the one
in `def_function` was renamed to start with underscore. It is a private
field, so it should be fine.

alextp reviewed (dismissed):
Thanks for catching this!

Can you add a unit test so this doesn't get dropped in the future?

foxik commented:
@alextp Thanks, I will add the test later. However, the builds are failing, because of the following reason:
- in `def_function.py` the field is called `experimental_relax_shapes`:
https://github.com/tensorflow/tensorflow/blob/9c5e6b7be87defd3ca7c43bff4e6f5ec79772f01/tensorflow/python/eager/def_function.py#L408
- however, in `function.py` it is called `_experimental_relax_shapes`:
https://github.com/tensorflow/tensorflow/blob/9c5e6b7be87defd3ca7c43bff4e6f5ec79772f01/tensorflow/python/eager/function.py#L2386

How do you think I should solve it? Something like testing for every one of them in turn and returning False if not defined, i.e., something like
```python
getattr(original_function, "_experimental_relax_shapes", getattr(original_function, "experimental_relax_shapes", False))
```
?
Or should one of the fields be renamed (but that seems like a large breaking change).

alextp commented:
Let's rename the fields as they are not on a public type.

On Wed, Dec 11, 2019 at 9:37 AM Milan Straka <notifications@github.com>
wrote:

> @alextp <https://github.com/alextp> Thanks, I will add the test later.
> However, the builds are failing, because of the following reason:
>
>    - in def_function.py the field is called experimental_relax_shapes:
>
>    https://github.com/tensorflow/tensorflow/blob/9c5e6b7be87defd3ca7c43bff4e6f5ec79772f01/tensorflow/python/eager/def_function.py#L408
>    - however, in function.py it is called _experimental_relax_shapes:
>
>    https://github.com/tensorflow/tensorflow/blob/9c5e6b7be87defd3ca7c43bff4e6f5ec79772f01/tensorflow/python/eager/function.py#L2386
>
> How do you think I should solve it? Something like testing for every one
> of them in turn and returning False if not defined, i.e., something like
>
> getattr(original_function, "_experimental_relax_shapes", getattr(original_function, "experimental_relax_shapes", False))
>
> ?
> Or should one of the fields be renamed (but that seems like a large
> breaking change).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35021?email_source=notifications&email_token=AAABHRK4WDXBLXA7K3AMOJLQYEQL5A5CNFSM4JZK4GR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGT6VVI#issuecomment-564652757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRLTCRJ4SFVCWACKIDDQYEQL5ANCNFSM4JZK4GRQ>
> .
>


-- 
 - Alex

foxik commented:
@alextp Thanks for guidance. I modified the commit to rename both fields to `_experimental_relax_shapes` which seems more consistent with the other fields. I also added a test. I was not sure if it is a good idea to access the `_experimental_relax_shapes` directly, so I reused the test for experimental_relax_shapes functionality (third call with a different shape should trigger a retrace with undefined batch size).

I have no way to run the tests on my home laptop, so I will wait for the results here, and fix if necessary.

foxik commented:
@alextp Thank you for your help. Would it make sense to CP this to 2.1? I met several people who thought `experimental_relax_shapes` does not work for them, because they added it to an instance method. The commit itself is a small and clean bugfix, so I think it could be added even this late in the release cycle.
